### ISSUE TITLE
Some spec elements received empty on handler when CR applied

### ISSUE THREAD

**Issue by neocorp** [nolar/kopf#555]
## Long story short

I have a CRD which has several fields with different data types. When I apply a CR for this CRD, kopf create handler receives and empty **groups** field in spec, same goes for body['spec'] as well.
## Description

When CR applied for the CRD, **groups** field that exists in the spec is received as an empty dict. While creating the CR, no validation errors occur so I believe groups data structure is fine. Below is the output for the spec received in which the **groups** is an empty dict:
```
Spec provided:{'groups': {}, 'namespace': 'testhosting', 'status': 'Initial', 'version': '1.0.0'}
```

<details><summary>The code snippet to reproduce the issue</summary>

**Here's the CRD:**
```apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: hosting.myproduct.com
spec:
  group: myproduct.com
  versions:
    - name: v1
      served: true
      storage: true
      schema:
        openAPIV3Schema:
          type: object
          properties:
            apiVersion:
              type: string
            spec:
              type: object
              properties:
                status:
                  type: string
                namespace:
                  type: string
                version:
                  type: string
                groups:
                  type: object
                  items:
                    type: array
                    items:
                      type: string
  scope: Namespaced
  names:
    kind: Hosting
    listKind: HostingList
    plural: hostings
    shortNames:
    - hst
    - hsts
    singular: hosting
```

**Here's the CR:**
```apiVersion: hosting.myproduct.com/v1
kind: Hosting
metadata:
  name: testhosting
spec:
  status: 'Initial'
  namespace: testhosting
  version: 1.0.0
  groups:
    devops: ["random.user@email.com", "another.user@email.com"]
    engineering: ["x.user@email.com", "y.user@email.com"]
    admin: ["admin@email.com", "admin.1@email.com"]
```

Handler:
```python
import kopf

@kopf.on.create('myproduct.com', 'v1', 'hostings')
def create_fn(spec, body, name, logger, **kwargs):
    spec_name = name
    spec_ns = spec['namespace']
    spec_groups = spec['groups']
    spec_body = spec['body']

    logger.info(f"Spec provided:{spec}") # groups field is empty
    logger.info(f"Name:{name}")
    logger.info(f"Namespace:{spec_ns}")
    logger.info(f"Spec Groups:{spec_groups}") # empty
    logger.info(f"Body:{spec_body}")   # empty
```
</details>

<details><summary>The exact command to reproduce the issue</summary>

```bash
kubectl apply -f crd.yaml
kopf run hostings.py --verbose
kubectl apply -f cr.yaml
```
</details>

<details><summary>The full output of the command that failed</summary>
There's no error produced however **body** is empty and **spec** is missing **groups** field data as shown below:
**Spec provided:{'groups': {}, 'namespace': 'someproduct', 'status': 'Initial', 'version': '1.0.0'}**
</details>

## Environment

* Kopf version: kopf, version 0.27
* Kubernetes version: 1.18
* Python version: 3.8.5
* OS/platform: MacOS

<details><summary>Python packages installed</summary>

```
aiohttp==3.6.2
aiojobs==0.2.2
async-timeout==3.0.1
attrs==20.1.0
avionix==0.3.1
boto3==1.14.54
botocore==1.17.54
cachetools==4.1.1
certifi==2020.6.20
chardet==3.0.4
click==7.1.2
docutils==0.15.2
gitdb==4.0.5
GitPython==3.1.7
google-auth==1.21.0
grpcio==1.31.0
grpcio-tools==1.31.0
idna==2.10
iso8601==0.1.12
jmespath==0.10.0
kopf==0.27
kubernetes==11.0.0
multidict==4.7.6
oauthlib==3.1.0
pip==20.1.1
protobuf==3.13.0
pyasn1==0.4.8
pyasn1-modules==0.2.8
pyhelm==2.14.5
pykube-ng==20.7.2
python-dateutil==2.8.1
PyYAML==5.3.1
requests==2.24.0
requests-oauthlib==1.3.0
rsa==4.6
s3transfer==0.3.3
setuptools==49.2.0
six==1.15.0
smmap==3.0.4
supermutes==0.2.5
typing-extensions==3.7.4.3
urllib3==1.25.10
websocket-client==0.57.0
wheel==0.34.2
yarl==1.5.1
```
</details>

**Comment by nolar** [comment]
Hello.

Does this happen when the operator is fully stopped, and you create the object?

It looks like a problem with the schema: you have `groups` defined as an object with only `groups.items` field, but attempt to store `groups.devops`, `groups.engineering`, etc. K8s removes the fields that do not match the schema. What's worse: it removes them silently.

This problem is known in Kopf, but it is not a problem of Kopf — it happens since K8s 1.16+ due to _"structural schemas"_. Some workarounds are documented here: 

* https://kopf.readthedocs.io/en/latest/configuration/#handling-progress — in the big blue "Note" section.
* https://kopf.readthedocs.io/en/latest/walkthrough/resources/?highlight=%22x-kubernetes-preserve-unknown-fields%22

TL;DR: You need `x-kubernetes-preserve-unknown-fields: true` on your spec to see where the fields are lost.

**Comment by neocorp** [comment]
Hi Nolar,

Thank you for the fast feedback. Just like you've suggested, I was able to figure out what was going on with `x-kubernetes-preserve-unknown-fields: true`. The problem was happening both when the operator was running and stopped. 

Thank you for helping out on this, we can close the issue.
